{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "76043bff-7d6e-4f75-980b-556360bac6b7",
   "metadata": {},
   "source": [
    "# Week 1: Sequence Analysis in Python\n",
    "\n",
    "## Overview\n",
    "This week focuses on implementing fundamental DNA sequence analysis in Python:\n",
    "- File parsing for FASTA format\n",
    "- Cleaning DNA sequences\n",
    "- Counting nucleotides\n",
    "- GC content calculation\n",
    "- Transcription (DNA -> RNA)\n",
    "- Complement and reverse complement\n",
    "- Start codon detection\n",
    "- Longest GC stretch\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "32ddea2d-1226-4ccc-afcf-1fb29e6b782a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Header: Human_sequence\n",
      "Base Count: Counter({'A': 9, 'T': 8, 'G': 8, 'C': 8})\n",
      "\n",
      "GC Content: 48.48%\n",
      "\n",
      "RNA: AUGCUAGCUAGCUAACGAUGCUAGCUAGCUGAC\n",
      "\n",
      "Complement Seq: TACGATCGATCGATTGCTACGATCGATCGACTG\n",
      "\n",
      "Reverse Complement: GTCAGCTAGCTAGCATCGTTAGCTAGCTAGCAT\n",
      "\n",
      "Position of start codon in Sequence: 0\n",
      "\n",
      "Longest GC Stretch: ('GC', 2)\n",
      "\n",
      "Header: Mouse_sequence\n",
      "Base Count: Counter({'T': 10, 'G': 10, 'C': 8, 'A': 8})\n",
      "\n",
      "GC Content: 50.0%\n",
      "\n",
      "RNA: UUGCGCGGAUCGUAGCUAGCUAGCUAGCUAAUGCUA\n",
      "\n",
      "Complement Seq: AACGCGCCTAGCATCGATCGATCGATCGATTACGAT\n",
      "\n",
      "Reverse Complement: TAGCATTAGCTAGCTAGCTAGCTACGATCCGCGCAA\n",
      "\n",
      "Position of start codon in Sequence: 30\n",
      "\n",
      "Longest GC Stretch: ('GCGCGG', 6)\n",
      "\n",
      "Header: Plant_sequence\n",
      "Base Count: Counter({'G': 8, 'C': 8, 'T': 8, 'A': 8})\n",
      "\n",
      "GC Content: 50.0%\n",
      "\n",
      "RNA: GCUAGCUAGCAUCGAUCGUAUAGCUAGCUAGC\n",
      "\n",
      "Complement Seq: CGATCGATCGTAGCTAGCATATCGATCGATCG\n",
      "\n",
      "Reverse Complement: GCTAGCTAGCTATACGATCGATGCTAGCTAGC\n",
      "\n",
      "Position of start codon in Sequence: None\n",
      "\n",
      "Longest GC Stretch: ('GC', 2)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# File loading and parsing\n",
    "def file_parsing(filepath):\n",
    "\n",
    "    fasta_dict = {}\n",
    "    with open(filepath, \"r\") as file:\n",
    "        header= None\n",
    "        seq = []\n",
    "        for line in file:\n",
    "            line= line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if header:\n",
    "                    fasta_dict[header] = \"\".join(seq)\n",
    "                header = line[1:]\n",
    "                seq= []\n",
    "            else:\n",
    "                seq.append(line)\n",
    "\n",
    "        if header:\n",
    "            fasta_dict[header] = \"\".join(seq)\n",
    "    return fasta_dict\n",
    "\n",
    "# seq cleaning \n",
    "def clean(seq):\n",
    "    clean_seq= \"\".join([base for base in seq.upper() if base in \"ATGC\"])\n",
    "    return clean_seq\n",
    "\n",
    "# Counter for counting the nucleotides counts\n",
    "from collections import Counter\n",
    "def counting(clean_seq):\n",
    "    base_count= Counter(clean_seq)\n",
    "    return base_count\n",
    "\n",
    "# GC Content determines the stability of DNA\n",
    "def calculate_gc(clean_seq):\n",
    "    gc = 0\n",
    "    for base in clean_seq:\n",
    "        if base in \"GC\":\n",
    "            gc +=1\n",
    "    gc_content= round((gc/ len(clean_seq)) * 100, 2)\n",
    "    return gc_content\n",
    "\n",
    "\n",
    "# Transcribes DNA -> RNA\n",
    "def transcription(clean_seq):\n",
    "    rna = \"\"\n",
    "    for base in clean_seq:\n",
    "        if base == \"T\":\n",
    "            rna += \"U\"\n",
    "        else:\n",
    "            rna += base\n",
    "    return rna\n",
    "\n",
    "\n",
    "# Finds the complement and reverse complement as the DNA structure \n",
    "# is anti-parallel\n",
    "def complmt(clean_seq):\n",
    "    complement= {\"A\":\"T\", \"T\":\"A\", \"G\":\"C\", \"C\":\"G\"}\n",
    "    complement_seq = \"\"\n",
    "    for base in clean_seq:\n",
    "        if base in complement:\n",
    "            complement_seq += complement[base]\n",
    "    reverse_complement= complement_seq[::-1]\n",
    "\n",
    "    return complement_seq, reverse_complement\n",
    "\n",
    "\n",
    "# Finds the start index of the start codon in a sequence\n",
    "def start_position(clean_seq):\n",
    "    start_codon = \"ATG\"\n",
    "    codons = [clean_seq[i:i+3] for i in range(0, len(clean_seq)-2, 3)]\n",
    "    if start_codon in codons:\n",
    "        codon_index = codons.index(start_codon) \n",
    "        start_index = codon_index * 3 \n",
    "        return start_index\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "# Finds the seq with the longest gc length and resets when the bases are AT\n",
    "def longest_gc_stretch(clean_seq):\n",
    "    current_length = 0\n",
    "    current_seq = \"\"\n",
    "    max_length = 0\n",
    "    max_seq = \"\"\n",
    "\n",
    "    for base in clean_seq:\n",
    "        if base in \"GC\":\n",
    "            current_length += 1\n",
    "            current_seq += base\n",
    "# This updates the max_length each time the gc sequence adds up\n",
    "            if current_length > max_length: \n",
    "                max_length = current_length\n",
    "                max_seq = current_seq\n",
    "\n",
    "# This resets the gc count once an A or T base is reached\n",
    "        else:\n",
    "            current_length = 0\n",
    "            current_seq = \"\"\n",
    "    return max_seq, max_length\n",
    "\n",
    "    \n",
    "sequences= file_parsing(\"cleaned_sequences.fasta\")\n",
    "\n",
    "\n",
    "for header, seq in sequences.items():\n",
    "    cleaned_seq= clean(seq)\n",
    "\n",
    "    print(f\"Header: {header}\")\n",
    "    \n",
    "    Base= counting(cleaned_seq)\n",
    "    print(f\"Base Count: {Base}\\n\")\n",
    "    \n",
    "    gc= calculate_gc(cleaned_seq)\n",
    "    print(f\"GC Content: {gc}%\\n\")\n",
    "    \n",
    "    Rna= transcription(cleaned_seq)\n",
    "    print(f\"RNA: {Rna}\\n\")\n",
    "    \n",
    "    compleme, Reverse = complmt(cleaned_seq)\n",
    "    print(f\"Complement Seq: {compleme}\\n\")\n",
    "    print(f\"Reverse Complement: {Reverse}\\n\")\n",
    "    \n",
    "    Pos= start_position(cleaned_seq)\n",
    "    print(f\"Position of start codon in Sequence: {Pos}\\n\")\n",
    "    \n",
    "    Longest_gc= longest_gc_stretch(cleaned_seq)\n",
    "    print(f\"Longest GC Stretch: {Longest_gc}\\n\")\n",
    "    \n",
    "                    \n",
    "               \n",
    "                    \n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ff3375-948c-44eb-95da-efdb7463e22f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
