{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cc937dae-a72f-4cbb-9aff-71a1dfa3bd1b",
   "metadata": {},
   "source": [
    "## WEEK 1 (August 5-11,2025): Bioinformatics File Formats + Python File Parsing\n",
    "\n",
    "Mini Project Questions: FASTA Sequence Analyzer\n",
    "\n",
    "Write a Python program to do the following:\n",
    "\n",
    "    Read a multi-FASTA file from disk (e.g., \"fasta.txt\")\n",
    "\n",
    "    Parse the FASTA file into a dictionary where keys are sequence headers and values are sequences\n",
    "\n",
    "    Clean sequences to include only valid DNA bases (A, T, G, C)\n",
    "\n",
    "    Calculate and print the length and GC content (%) of each sequence\n",
    "\n",
    "    Calculate and print the average length of all sequences combined\n",
    "\n",
    "    Find and print the header, length, and GC content of the longest sequence\n",
    "\n",
    "Bonus:\n",
    "\n",
    "    Print the actual nucleotide sequence of the longest sequence (optional)\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "65fa89cd-294f-462f-8997-8d57de316c94",
   "metadata": {},
   "source": [
    "# A FASTA file is a file that stores one or more biological sequence, like this\n",
    ">sequence1\n",
    "ATTTTGGGCTCTTCTCGGG\n",
    ">sequence2\n",
    "AAAAATTTGGCCTGGTTTTTT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "be525a8c-9326-461b-8067-35a7a49ed53e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastafile= \"\"\"\n",
    ">seq1 Homo sapiens gene A\n",
    "ATGCGTAGCTAGTCGATCGATCGATCGATCGA\n",
    ">seq2 Mus musculus gene B\n",
    "GGCGGCGTTAGCTAGCTAGGCTAGCTAGCTAG\"\"\"\n",
    "\n",
    "\n",
    "with open(\"sample.fasta\", \"w\") as file:\n",
    "    file.write(fastafile)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0562d4bd-99e3-40e6-b2f3-3f6a72cd3221",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "with open(\"sample.fasta\", \"r\") as file:\n",
    "    header_count = 0\n",
    "    for line in file:\n",
    "        line= line.strip()\n",
    "        if line.startswith('>'):\n",
    "            header_count +=1 \n",
    "    print(header_count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "54f9a4d4-2bdc-4ade-8c56-17e244c27292",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna= \"\"\"\n",
    ">sequence_1\n",
    "ATGCTAGCTAGCTACGATCGATCGTAGCTAGCTAGCATCG\n",
    "ATCGATGCTAGCTAGCTAGCATCGATGCTAGCTAGCATCG\n",
    ">sequence_2\n",
    "GGCATCGATCGATCGATCGATCGGATCGATCGATCGATCG\n",
    "CGATCGATCGATCGATCGATCGGATCGATCGATCGATCGA\"\"\"\n",
    "\n",
    "with open(\"fasta.txt\", \"w\") as file:\n",
    "    file.write(dna.strip())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f214729b-a667-45db-946a-7faa62afd7e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence_1 - Length: 80 - GC%: 50.0\n",
      "sequence_2 - Length: 80 - GC%: 53.75\n"
     ]
    }
   ],
   "source": [
    "def fasta_seq(fastatxt):\n",
    "    fasta_dict= {}\n",
    "    with open(\"fasta.txt\", \"r\") as file:\n",
    "        header= None\n",
    "        sequence_lines= []\n",
    "        for line in file:\n",
    "            line= line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if header:\n",
    "                    fasta_dict[header]= \"\".join(sequence_lines)\n",
    "                header= line[1:]\n",
    "                sequence_lines= []\n",
    "\n",
    "            else:\n",
    "                sequence_lines.append(line)\n",
    "        if header:\n",
    "            fasta_dict[header]= \"\".join(sequence_lines)\n",
    "    return fasta_dict\n",
    "\n",
    "def clean(sequence_lines):\n",
    "    cleaned_sequences= \"\".join([base for base in sequence_lines if base in 'ACGT'])\n",
    "    return cleaned_sequences\n",
    "\n",
    "def calculate_gc(cleaned_sequences):\n",
    "    gc=0\n",
    "    for base in cleaned_sequences:\n",
    "        if base in 'GC':\n",
    "            gc += 1\n",
    "    gc_percent = round(gc/ len(cleaned_sequences) * 100, 2)\n",
    "    return gc_percent\n",
    "\n",
    "sequences= fasta_seq(\"fasta.txt\")\n",
    "\n",
    "for header, seq in sequences.items():\n",
    "    cleaned= clean(seq)\n",
    "    gc= calculate_gc(cleaned)\n",
    "    print(f\"{header} - Length: {len(cleaned)} - GC%: {gc}\")  \n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5452d5cb-2e4f-456f-a7db-d5fabd27d15f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence_1 - Length: 80 - GC%: 50.0\n",
      "sequence_2 - Length: 80 - GC%: 53.75\n",
      "Average Sequence Length:80.0\n",
      "Longest Sequence: \n",
      "Header: sequence_2\n",
      "Length: 80\n",
      "GC%: 53.75\n"
     ]
    }
   ],
   "source": [
    "def fasta_seq(fastatxt):\n",
    "    fasta_dict= {}  # dictionary to store the header and seqs merged into a single sequence\n",
    "    with open(\"fasta.txt\", \"r\") as file:\n",
    "        header= None  # this shows that no previous header before sequence_1\n",
    "        sequence_lines= [] # this is for merging/appending the sequences belonging to a particular \n",
    "                            # header into a single sequence\n",
    "        for line in file:\n",
    "            line= line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if header: # this answers the question if there is a previous header known\n",
    "                    fasta_dict[header]= \"\".join(sequence_lines) # this answers what to do to the previous\n",
    "                                                                # header which is to save it\n",
    "                header= line[1:]  # this step tells what to do if there is None\n",
    "                sequence_lines= []\n",
    "            else:\n",
    "                sequence_lines.append(line) \n",
    "\n",
    "        if header:\n",
    "            fasta_dict[header]= \"\".join(sequence_lines) # this steps shows how to add the last header not \n",
    "                                                        # to be omitted\n",
    "    return fasta_dict\n",
    "\n",
    "def clean(sequence_lines):\n",
    "    cleaned_sequences= \"\".join([base for base in sequence_lines if base in 'ATGC'])\n",
    "    return cleaned_sequences\n",
    "\n",
    "def calculate_gc(cleaned_sequences):\n",
    "    gc= 0\n",
    "    for base in cleaned_sequences.upper():\n",
    "        if base in 'GC':\n",
    "            gc +=1\n",
    "    gc_percent= round(gc/len(cleaned_sequences) *100, 2)\n",
    "    return gc_percent\n",
    "\n",
    "def avg_length(fasta_dict):\n",
    "    lengths= []\n",
    "    for each_sequence in fasta_dict.values():\n",
    "        lengths.append(len(each_sequence))\n",
    "    average= round(sum(lengths)/len(lengths), 2)\n",
    "    return average\n",
    "\n",
    "def find_longest_sequence(fasta_dict):\n",
    "    longest_sequence= \"\"\n",
    "    max_length = 0\n",
    "    for header, seq in fasta_dict.items():\n",
    "        if len(seq)> max_length:\n",
    "            max_length = len(seq)\n",
    "            longest_sequence = header\n",
    "          \n",
    "    return longest_sequence, max_length\n",
    "        \n",
    "    \n",
    "\n",
    "sequences= fasta_seq(\"fasta.txt\")\n",
    "\n",
    "for header, seq in sequences.items():\n",
    "    cleaned= clean(seq)\n",
    "    gc = calculate_gc(cleaned)\n",
    "    print(f\"{header} - Length: {len(cleaned)} - GC%: {gc}\")\n",
    "\n",
    "\n",
    "average = avg_length(sequences)\n",
    "print(f\"Average Sequence Length:{average}\")\n",
    "\n",
    "longest_sequence, max_length= find_longest_sequence(sequences)\n",
    "gc = calculate_gc(cleaned)\n",
    "print(f\"Longest Sequence: \\nHeader: {header}\\nLength: {max_length}\\nGC%: {gc}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc69bf65-c838-45ab-9af7-137b6a8cf5e0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
