{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6cda8a87-0cb5-4e55-9bc7-015cfab3f2d6",
   "metadata": {},
   "source": [
    "# DNA & RNA Sequence Manipulation Exercises\n",
    "\n",
    "This notebook contains exercises on:\n",
    "- Reversing DNA sequences\n",
    "- Finding reverse complements\n",
    "- Counting codon occurrences (e.g., ATG)\n",
    "- Sliding window analysis\n",
    "- Replacing nucleotides in RNA/DNA\n",
    "- Removing special characters\n",
    "- Splitting multifasta strings\n",
    "- Motif position identification"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efecb9f3-16d4-4bd9-a660-59c85c90dffe",
   "metadata": {},
   "source": [
    "### Python & Bioinformatics Exercises Summary â€” 1-7th July 2025\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "01c6b0f5-3442-475d-abd6-32dcfce28df7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a dna sequence ACCCTTTGGG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGGTTTCCCA\n"
     ]
    }
   ],
   "source": [
    "# Exercise 1: Reverse a DNA sequence\n",
    "\n",
    "seq= input(\"Enter a dna sequence\").upper()\n",
    "reverse = seq[::-1]\n",
    "print(reverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7c7fd31f-ff99-49e9-a24e-632cdc210fae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a DNA sequence agggttttccc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCCCAAAAGGG\n"
     ]
    }
   ],
   "source": [
    "# Exercise 2: Get the reverse complement of a DNA sequence\n",
    "seq = input(\"Enter a DNA sequence\").upper()\n",
    "complement = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "result = ''\n",
    "for base in seq:\n",
    "    if base in complement:\n",
    "        result += complement[base]\n",
    "    else:\n",
    "        result += base\n",
    "print(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "577de892-8f84-41f9-a18c-d5870d113125",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a dna sequence aaaatgggggagtttagttttcccc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "# Exercise 3: Count how many times \"ATG\" appears\n",
    "seq= input(\"Enter a dna sequence\").upper()\n",
    "count= seq.count('ATG')\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2c3d503b-f515-4af5-91d7-4ec04ba01eb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter an dna sequence: aaatgatgatgggggtttcccatg\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count of atg :1\n"
     ]
    }
   ],
   "source": [
    "# Another method (IN-FRAME COUNTING)\n",
    "seq= input(\"Enter an dna sequence:\").upper()\n",
    "count_atg = 0\n",
    "codons= [seq[i:i+3] for i in range(0, len(seq)-2,3) if len(seq[i:i+3])==3]\n",
    "for codon in codons:\n",
    "    if codon == 'ATG':\n",
    "        count_atg += 1\n",
    "print(f\"count of atg :{count_atg}\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a8467b74-0c3e-4d47-aa67-56026ceeb0d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a dna sequence: aaatgatgatgggggtttcccatg\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n"
     ]
    }
   ],
   "source": [
    "# Another method (SLIDING WINDOW/ OVERLAPPING/ ALL MATCHING)\n",
    "seq= input(\"Enter a dna sequence:\").upper()\n",
    "count_atg = 0\n",
    "motif = \"ATG\"\n",
    "for i in range(len(seq)- len(motif) + 1):\n",
    "    if seq[i:i+ len(motif)]== motif:\n",
    "        count_atg += 1\n",
    "print(count_atg)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a1560bcb-dbd9-4adb-81ee-fc0624c73926",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a rna sequence: auuuggggcccaaaugc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATTTGGGGCCCAAATGC\n"
     ]
    }
   ],
   "source": [
    "# Exercise 4: Replace all \"U\" with \"T\" (RNA-DNA)\n",
    "rna= input(\"Enter a rna sequence:\").upper()\n",
    "dna = ''\n",
    "for base in rna:\n",
    "    if base == 'U':\n",
    "        dna += 'T'\n",
    "    else:\n",
    "        dna += base\n",
    "print(dna)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f075e978-1ea4-4a21-b069-e3340c7e8c44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">seq1alpha\n"
     ]
    }
   ],
   "source": [
    "# Exercise 5: Remove special characters from sequence header\n",
    "# Task: Given a FASTA header string (which should only have letters, numbers,_and the starting >),\n",
    "# remove any unwanted special characters like #, *, @, etc\n",
    "\n",
    "header = \">seq#1*alpha\"\n",
    "clean_header = header.replace('#', '').replace('*', '')\n",
    "print(clean_header)\n",
    "                                                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0057cd2e-d07e-4df1-b244-afc814020242",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>seq1', 'ATGC', '>seq2', 'GGA']\n"
     ]
    }
   ],
   "source": [
    "# Exercise 6: Split a multi-FASTA string into lines\n",
    "# Task : Given a multi-FASTA formatted string with sequences and headers separated by newlines, split\n",
    "# it into a list of lines.\n",
    "\n",
    "fasta= \">seq1\\nATGC\\n>seq2\\nGGA\"\n",
    "splitted = fasta.splitlines()\n",
    "print(splitted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79f64ba6-1c26-4407-ada4-df2ec4fa64b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>seq1', 'ATGC', '>seq2', 'GGA']\n"
     ]
    }
   ],
   "source": [
    "# Another method for Exercise 6\n",
    "\n",
    "fasta= \">seq1\\nATGC\\n>seq2\\nGGA\"\n",
    "splitted = fasta.split(\"\\n\")\n",
    "print(splitted)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "65a9bd65-e53e-4937-b310-9cde3c58eba2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a dna sequence: aaacgcgcgcgccgcgcg\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3, 5, 7, 9, 12, 14, 16]\n"
     ]
    }
   ],
   "source": [
    "# Exercise 7: Find all positions of motif \"CG\" \n",
    "# Task: Given a DNA sequence, find all positions where the motif \"CG\" occurs.\n",
    "# Positions are based on the first character of each match\n",
    "# Return them as a list of numbers.\n",
    "\n",
    "seq= input(\"Enter a dna sequence:\").upper()\n",
    "motif= \"CG\"\n",
    "positions = []\n",
    "for i in range(len(seq) - len(motif) + 1):\n",
    "    if seq[i:i+len(motif)]== motif:\n",
    "        positions.append(i)\n",
    "\n",
    "print(positions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a296222a-63fc-45c9-ba12-2467a5c0a863",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a sequence: aaatttgggggggtga\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "# Exercise 8: Check if sequence starts with \"ATG\" and ends with a stop codon\n",
    "# Task:\n",
    "# write code to check if:\n",
    "# The sequence starts with \"ATG\" (start codon)\n",
    "# The sequence ends with any one of these stop codons: \"TAA\", \"TAG\", or \"TGA\"\n",
    "# if both conditions are true, print True; else print False\n",
    "\n",
    "seq= input(\"Enter a sequence:\").upper()\n",
    "           \n",
    "start_codon = \"ATG\"\n",
    "stop_codon = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "\n",
    "start = seq.startswith(start_codon)\n",
    "stop= seq.endswith(tuple(stop_codon))\n",
    "\n",
    "if start and end:\n",
    "    print(True)\n",
    "else:\n",
    "    print(False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8880a992-a06a-40db-9bf6-3ccdda323fd2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
