{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "95986137-3869-465f-a01f-a6ec735807f9",
   "metadata": {},
   "source": [
    "# Translating Mutated RNA to Protein\n",
    "\n",
    "## Description\n",
    "This notebook shows:\n",
    "- Translating RNA sequences from files into proteins\n",
    "- Handling mutated sequences\n",
    "- Applying Python string manipulation and bioinformatics translation methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1ea486b-a1a4-4fcb-bd86-3351fe15b6a6",
   "metadata": {},
   "source": [
    "Challenge: Translate mutated RNA sequence to protein\n",
    "\n",
    "Task:\n",
    "\n",
    "Read a DNA sequence from a file (like before)\n",
    "\n",
    "Clean it (only A, T, G, C)\n",
    "\n",
    "Apply the same mutation (A→G, T→C)\n",
    "\n",
    "Transcribe the mutated DNA to RNA (replace T with U)\n",
    "\n",
    "Translate the RNA sequence into a protein sequence using the standard codon table\n",
    "\n",
    "Stop translation at the first stop codon\n",
    "\n",
    "Output the mutated DNA, RNA, and translated protein sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f94df66f-4979-4b30-a2ec-9ff5cbdedcac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutated DNA:GGCCGGGCGCGCCGGGCGC\n",
      "RNA: GGCCGGGCGCGCCGGGCGC\n",
      "Translated Protein: GRARRA\n"
     ]
    }
   ],
   "source": [
    "with open(\"dna.txt\", \"r\") as file:\n",
    "    sequence= file.read().upper()\n",
    "\n",
    "cleaned_sequence= \"\".join([base for base in sequence if base in 'ATGC'])\n",
    "mutated_dna= \"\"\n",
    "for base in cleaned_sequence:\n",
    "    if base == 'A':\n",
    "        mutated_dna +='G'\n",
    "    elif base == 'T':\n",
    "        mutated_dna +='C'\n",
    "    else:\n",
    "        mutated_dna += base\n",
    "\n",
    "codon_table = {\n",
    "    'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L',\n",
    "    'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L',\n",
    "    'AUU':'I', 'AUC':'I', 'AUA':'I', 'AUG':'M',\n",
    "    'GUU':'V', 'GUC':'V', 'GUA':'V', 'GUG':'V',\n",
    "    'UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S',\n",
    "    'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "    'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "    'GCU':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A',\n",
    "    'UAU':'Y', 'UAC':'Y', 'UAA':'_', 'UAG':'_',\n",
    "    'CAU':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'UGU':'C', 'UGC':'C', 'UGA':'_', 'UGG':'W',\n",
    "    'CGU':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R',\n",
    "    'AGU':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'GGU':'G', 'GGG':'G', 'GGC':'G', 'GGA':'G'\n",
    "}\n",
    "rna= \"\"\n",
    "for base in mutated_dna:\n",
    "    if base == 'T':\n",
    "        rna +='U'\n",
    "    else:\n",
    "        rna += base\n",
    "\n",
    "codons= [rna[i:i+3] for i in range (0, len(rna)-2, 3)]\n",
    "aminoacids= list(map(lambda codon:codon_table.get(codon, '?'), codons))\n",
    "\n",
    "translated_protein= []\n",
    "for aa in aminoacids:\n",
    "    if aa == '_':\n",
    "        break\n",
    "    else:\n",
    "        translated_protein.append(aa)\n",
    "\n",
    "print(f\"Mutated DNA:{mutated_dna}\\nRNA: {rna}\\nTranslated Protein: {\"\".join(translated_protein)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8e84bf86-06c0-44ef-8041-c3e10aa46f78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutated DNA: GGCCGGGCGCGCCGGGCGC\n",
      "RNA: GGCCGGGCGCGCCGGGCGC\n",
      "TranslatedProteins: GRARRA\n"
     ]
    }
   ],
   "source": [
    "# Another method\n",
    "with open(\"dna.txt\", \"r\") as file:\n",
    "    sequence= file.read().upper()\n",
    "    \n",
    "cleaned_sequence= \"\".join([base for base in sequence if base in 'ATGC'])\n",
    "\n",
    "mutated_complement= {'A':'G', 'T':'C'}\n",
    "mutated_dna= \"\"\n",
    "for base in cleaned_sequence:\n",
    "    if base in mutated_complement:\n",
    "        mutated_dna += mutated_complement[base]\n",
    "    else:\n",
    "        mutated_dna += base\n",
    "\n",
    "rna= mutated_dna.replace('T', 'U')\n",
    "\n",
    "codon_table = {\n",
    "    'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L',\n",
    "    'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L',\n",
    "    'AUU':'I', 'AUC':'I', 'AUA':'I', 'AUG':'M',\n",
    "    'GUU':'V', 'GUC':'V', 'GUA':'V', 'GUG':'V',\n",
    "    'UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S',\n",
    "    'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "    'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "    'GCU':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A',\n",
    "    'UAU':'Y', 'UAC':'Y', 'UAA':'_', 'UAG':'_',\n",
    "    'CAU':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'UGU':'C', 'UGC':'C', 'UGA':'_', 'UGG':'W',\n",
    "    'CGU':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R',\n",
    "    'AGU':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'GGU':'G', 'GGG':'G', 'GGC':'G', 'GGA':'G'\n",
    "}\n",
    "\n",
    "codons= [rna[i:i+3] for i in range(0, len(rna)-2, 3)]\n",
    "aminoacids= []\n",
    "for codon in codons:\n",
    "    if codon in codon_table:\n",
    "        aa= codon_table[codon]\n",
    "        if aa == '_':\n",
    "            break\n",
    "        aminoacids.append(aa)\n",
    "\n",
    "print(f\"Mutated DNA: {mutated_dna}\\nRNA: {rna}\\nTranslatedProteins: {\"\".join(aminoacids)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e730f4f-3e1c-4608-be72-47dc0ab9f50a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
