{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dc7fd7b2-1727-48c8-91f4-bdf8880f119a",
   "metadata": {},
   "source": [
    "# Compare Mutated vs Wild Type Protein\n",
    "\n",
    "## Description\n",
    "This notebook demonstrates:\n",
    "- Comparing protein sequences from mutated and wild type RNA\n",
    "- Detecting differences in amino acid sequences\n",
    "- Using Python for sequence comparison and basic bioinformatics analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97dc4a8b-4482-4a07-bace-732245afa8ed",
   "metadata": {},
   "source": [
    "Challenge: Compare Mutated Protein to Wild-Type Protein\n",
    "\n",
    "Task:\n",
    "\n",
    "You already have the mutated DNA → RNA → protein sequence.\n",
    "\n",
    "Now:\n",
    "\n",
    "1. Also translate the original (wild-type) DNA into a protein.\n",
    "\n",
    "\n",
    "2. Compare the wild-type and mutated proteins.\n",
    "\n",
    "\n",
    "3. Identify differences (amino acid substitutions).\n",
    "\n",
    "\n",
    "4. Print both protein sequences + indicate where the mutation(s) occurred.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fa954f91-d9b3-4a7a-b348-2606866f0817",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wild-type protein: TFSALD\n",
      "Mutated protein: APGAPG\n"
     ]
    }
   ],
   "source": [
    "with open(\"dna.txt\", \"r\") as file:\n",
    "    sequence= file.read().upper()\n",
    "\n",
    "cleaned_sequence= \"\".join([base for base in sequence if base in 'ATGC'])\n",
    "\n",
    "mutated_dna= \"\"\n",
    "mutation_complement= {'A':'G', 'T':'C'}\n",
    "for base in cleaned_sequence:\n",
    "    if base in mutation_complement:\n",
    "        mutated_dna += mutation_complement[base]\n",
    "    else:\n",
    "        mutated_dna += base\n",
    "\n",
    "complement_of_mutated_dna = {'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "mutated_dna_complement= \"\"\n",
    "for base in mutated_dna:\n",
    "    if base in complement_of_mutated_dna:\n",
    "        mutated_dna_complement += complement_of_mutated_dna[base]\n",
    "    else:\n",
    "        mutated_dna_complement += base\n",
    "\n",
    "reverse_complement_of_mutated_dna= mutated_dna_complement[::-1]\n",
    "\n",
    "rna= reverse_complement_of_mutated_dna.replace('T', 'U')\n",
    "\n",
    "codon_table = {\n",
    "    'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L',\n",
    "    'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L',\n",
    "    'AUU':'I', 'AUC':'I', 'AUA':'I', 'AUG':'M',\n",
    "    'GUU':'V', 'GUC':'V', 'GUA':'V', 'GUG':'V',\n",
    "    'UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S',\n",
    "    'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "    'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "    'GCU':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A',\n",
    "    'UAU':'Y', 'UAC':'Y', 'UAA':'_', 'UAG':'_',\n",
    "    'CAU':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'UGU':'C', 'UGC':'C', 'UGA':'_', 'UGG':'W',\n",
    "    'CGU':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R',\n",
    "    'AGU':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'GGU':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G'\n",
    "}\n",
    "\n",
    "translated = []\n",
    "codons= [rna[i:i+3] for i in range (0, len(rna)-2, 3)]\n",
    "aminoacids= list(map(lambda codon: codon_table.get(codon, '?'), codons))\n",
    "for aa in aminoacids:\n",
    "    if aa == '_':\n",
    "        break\n",
    "    elif aa == '?':\n",
    "        print(f\"Unknown codon found: {codon}\")\n",
    "    else:\n",
    "        translated.append(aa)\n",
    "\n",
    "mutated_protein= \"\".join(translated)\n",
    "\n",
    "# Translate the original(wild-type) DNA into protein\n",
    "\n",
    "complement= {'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "dna_complement= \"\"\n",
    "\n",
    "for base in cleaned_sequence:\n",
    "    if base in complement:\n",
    "        dna_complement += complement[base]\n",
    "    else:\n",
    "        dna_complement +=base\n",
    "\n",
    "reverse_complement = dna_complement[::-1]\n",
    "\n",
    "original_rna = reverse_complement.replace('T', 'U')\n",
    "\n",
    "original_codons= [original_rna[i:i+3] for i in range(0, len(original_rna)-2, 3)]\n",
    "original_translated=[]\n",
    "\n",
    "for codon in original_codons:\n",
    "    aa = codon_table.get(codon,'?')\n",
    "    if aa =='_':\n",
    "        break\n",
    "    elif aa =='?':\n",
    "        print(f\"Unknown codon found!!!: {codon}\")\n",
    "    else:\n",
    "        original_translated.append(aa)\n",
    "\n",
    "original_protein= \"\".join(original_translated)\n",
    "    \n",
    "print(f\"Wild-type protein: {original_protein}\\nMutated protein: {mutated_protein}\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a712b89-2c16-4fa5-859b-8fa7df4bc614",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
