{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b31f19fd-dd5b-4cdf-91e3-b105962dba2d",
   "metadata": {},
   "source": [
    "# Simulating Point Mutation and Translating Longest Protein\n",
    "\n",
    "## Description\n",
    "This notebook demonstrates:\n",
    "- Introducing point mutations into DNA sequences\n",
    "- Translating the mutated sequence to protein\n",
    "- Finding the longest open reading frame\n",
    "- Combining file handling and bioinformatics operations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6ee0e3-dad7-416d-9dee-5ae31c5a2821",
   "metadata": {},
   "source": [
    "Challenge: Simulate Point Mutation and Translate the Longest Protein\n",
    "\n",
    "Problem Statement:\n",
    "You have a DNA sequence stored in dna.txt.\n",
    "Your task is to:\n",
    "\n",
    "1. Clean the sequence to only valid bases: A, T, G, C\n",
    "\n",
    "\n",
    "2. Simulate a point mutation:\n",
    "\n",
    "Replace all 'A' with 'G'\n",
    "\n",
    "Replace all 'T' with 'C'\n",
    "\n",
    "\n",
    "\n",
    "3. Find the complement of the mutated DNA\n",
    "\n",
    "\n",
    "4. Find the reverse of that complement\n",
    "\n",
    "\n",
    "5. Transcribe the reversed complement into RNA (T → U)\n",
    "\n",
    "\n",
    "6. Split the RNA into codons\n",
    "\n",
    "\n",
    "7. Translate into amino acids using the standard RNA codon table\n",
    "\n",
    "\n",
    "8. Extract the longest continuous protein (no stop codons) from the start of the sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "513d1876-6361-46ec-b98e-ae744a836ede",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutated DNA: GGCCGGGCGCGCCGGGCGC\n",
      "Complement: CCGGCCCGCGCGGCCCGCG\n",
      "      \n",
      "Reverse Complement: GCGCCCGGCGCGCCCGGCC\n",
      "       \n",
      "RNA: GCGCCCGGCGCGCCCGGCC\n",
      "Codons: ['GCG', 'CCC', 'GGC', 'GCG', 'CCC', 'GGC']\n",
      "Longest Translated Protein: APGAPG\n"
     ]
    }
   ],
   "source": [
    "with open(\"dna.txt\", \"r\") as file:\n",
    "    sequence= file.read().upper()\n",
    "\n",
    "cleaned_sequence= \"\".join([base for base in sequence if base in 'ATGC'])\n",
    "\n",
    "mutation_bases= {'A':'G', 'T':'C'}\n",
    "mutated_dna = \"\"\n",
    "for base in cleaned_sequence:\n",
    "    if base in mutation_bases:\n",
    "        mutated_dna += mutation_bases[base]\n",
    "    else:\n",
    "        mutated_dna += base\n",
    "\n",
    "complement_of_mutated_dna =\"\"\n",
    "complement_bases= {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "for base in mutated_dna:\n",
    "    if base in complement_bases:\n",
    "        complement_of_mutated_dna += complement_bases[base]\n",
    "    else:\n",
    "        complement_of_mutated_dna += base\n",
    "\n",
    "reverse_complement_of_mutated_dna = complement_of_mutated_dna[::-1]\n",
    "\n",
    "rna= reverse_complement_of_mutated_dna.replace('T', 'U')\n",
    "\n",
    "codons= [rna[i:i+3] for i in range(0, len(rna)-2, 3)]\n",
    "\n",
    "codon_table = {\n",
    "    'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L',\n",
    "    'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L',\n",
    "    'AUU':'I', 'AUC':'I', 'AUA':'I', 'AUG':'M',\n",
    "    'GUU':'V', 'GUC':'V', 'GUA':'V', 'GUG':'V',\n",
    "    'UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S',\n",
    "    'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "    'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "    'GCU':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A',\n",
    "    'UAU':'Y', 'UAC':'Y', 'UAA':'_', 'UAG':'_',\n",
    "    'CAU':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'UGU':'C', 'UGC':'C', 'UGA':'_', 'UGG':'W',\n",
    "    'CGU':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R',\n",
    "    'AGU':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'GGU':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G'\n",
    "}\n",
    "\n",
    "aminoacids= list(map(lambda codon: codon_table.get(codon, '?'), codons))\n",
    "protein = []\n",
    "for aa in aminoacids:\n",
    "    if aa == '_':\n",
    "        break\n",
    "    elif aa =='?':\n",
    "        \"Unknown codon found\"\n",
    "    else:\n",
    "        protein.append(aa)\n",
    "\n",
    "longest_protein = \"\".join(protein)\n",
    "\n",
    "print(f\"\"\"Mutated DNA: {mutated_dna}\\nComplement: {complement_of_mutated_dna}\n",
    "      \\nReverse Complement: {reverse_complement_of_mutated_dna}\n",
    "       \\nRNA: {rna}\\nCodons: {codons}\\nLongest Translated Protein: {longest_protein}\"\"\")\n",
    "    \n",
    "                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcce8c2c-bf26-4c3b-ac02-c8d5ac447216",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "894d5952-2a28-4fb5-ae4a-c73506482d26",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b11f2a76-f50a-40b2-8505-e6a5bbd31730",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
